[runtimes] Append `-nostd*++` flags only for Clang

mgorny opened the pull request:
Append `-nostdlib++` and `-nostdinc++` flags to `CMAKE_REQUIRED_FLAGS`
only if we are actually building with Clang.  These flags are also
passed to the C compiler, which is not allowed in GCC.  Since CMake
implicitly performs some tests using the C compiler, this can lead
to incorrect check results.  This should be safe, since FWIU we only
need them when bootstrapping Clang.

Even though we know that Clang supports these flags, we still need
to explicitly check if they work, as in some scenarios adding
`-nostdlib++` actually breaks the build.  See PR #108357 for examples
of that.

Fixes #90332

jhuber6 reviewed (commented):
What do we need the required flags for again? I'm guessing this is just for checking C/C++ flags?

Meinersbur commented:
I think it was added because originally, LLVM_ENABLE_RUNTIMES was invented for libc++(abi). It is C++, but obviously cannot link against libc++.so before libc++.so is built. For any library that is not libc++ but uses the C++ standard library (flang-rt, hwloc through openmp, maybe others), as long as there is a libc++/libstdc++ installed on the system and found by Clang, it can use that.

If we want those runtimes to link together with just-built libc++, we need a 4-stage bootstrapping build (clang -> compiler-rt builtins -> libc++ -> flang-rt/openmp). Or, at least this seems to be the current approach, make the configure step not depend on the existence of the C++ standard library, but for build-time, one can add a dependency to libc++ if built in the same runtimes build.

Also see #131010

Note that you can temporarily get rid of the flags using [`cmake_push_check_state(RESET)`]().

Also consider that you can add flags only for a specific language: `$<$<COMPILE_LANGUAGE:C>:-nostdlib++>`. Haven't tries whether this also works for `CMAKE_REQUIRED_FLAGS`.

mgorny committed:
[runtimes] Append `-nostd*++` flags only for Clang

Append `-nostdlib++` and `-nostdinc++` flags to `CMAKE_REQUIRED_FLAGS`
only if we are actually building with Clang.  These flags are also
passed to the C compiler, which is not allowed in GCC.  Since CMake
implicitly performs some tests using the C compiler, this can lead
to incorrect check results.  This should be safe, since FWIU we only
need them when bootstrapping Clang.

Even though we know that Clang supports these flags, we still need
to explicitly check if they work, as in some scenarios adding
`-nostdlib++` actually breaks the build.  See PR #108357 for examples
of that.

Fixes #90332

Signed-off-by: Michał Górny <mgorny@gentoo.org>

jhuber6 commented:
> I think it was added because originally, LLVM_ENABLE_RUNTIMES was invented for libc++(abi). It is C++, but obviously cannot link against libc++.so before libc++.so is built. For any library that is not libc++ but uses the C++ standard library (flang-rt, hwloc through openmp, maybe others), as long as there is a libc++/libstdc++ installed on the system and found by Clang, it can use that.
> 
> If we want those runtimes to link together with just-built libc++, we need a 4-stage bootstrapping build (clang -> compiler-rt builtins -> libc++ -> flang-rt/openmp). Or, at least this seems to be the current approach, make the configure step not depend on the existence of the C++ standard library, but for build-time, one can add a dependency to libc++ if built in the same runtimes build.

I don't think we fully need to bootstrap it, we already have `libc++` built on top of `libc` and I built `flang-rt` on top of `libc++` built on to of `libc`. Shouldn't these just be a list of CMake dependencies? The runtimes order is deterministic so we just move those libraries earlier in the chain. Having the builtins separate only makes sense because oftentimes those are necessary for passing some feature checks by the compiler.

Meinersbur commented:
> The runtimes order is deterministic so we just move those libraries earlier in the chain. Having the builtins separate only makes sense because oftentimes those are necessary for passing some feature checks by the compiler.

And CMAKE_REQUIRED_FLAGS, that this PR is about, is used for feature checks. When building `runtimes-configure`, the order in the chain is irrelevant, since libc++.so has not been built in any case.

Consider my previous statement:

>  Or, at least this seems to be the current approach, make the configure step not depend on the existence of the C++ standard library

thesamesam reviewed (approved):
It looks fine (enough) for me if others are happy.

jhuber6 reviewed (approved):
Seems innocuous enough.
